cpp/src/strings/json/json_path.cu

```diff
@@ -0,0 +1,983 @@
+            if (idx - 1 != op.index) { return parse_result::ERROR; }
+            push_context(ctx.j_state, ctx.commands + 1, ctx.list_element);
+          }
+        }
+      } break;
+
+      // some sort of error.
+      case path_operator_type::ERROR: return parse_result::ERROR; break;
+
+      // END case
+      default: {
+        if (ctx.list_element && element_count > 0) {
+          output.add_output({"," DEBUG_NEWLINE, 1 + DEBUG_NEWLINE_LEN});
+        }
+        PARSE_TRY(ctx.j_state.extract_element(&output, ctx.list_element));
+        if (ctx.list_element && last_result != parse_result::EMPTY) { element_count++; }
+      } break;
+    }
+  }
+
+  return parse_result::SUCCESS;
+}
+
+// hardcoding this for now. to reach a stack depth of 8 would require
+// a jsonpath containing 7 nested wildcards so this is probably reasonable.
+constexpr int max_command_stack_depth = 8;
+
+/**
+ * @brief Parse a single json string using the provided command buffer
+ *
+ * This function exists primarily as a shim for debugging purposes.
+ *
+ * @param input The incoming json string
+ * @param input_len Size of the incoming json string
+ * @param commands The command buffer to be applied to the string. Always ends with a
+ * path_operator_type::END
+ * @param out_buf Buffer user to store the results of the query (nullptr in the size computation
+ * step)
+ * @param out_buf_size Size of the output buffer
+ * @returns A pair containing the result code the output buffer.
+ */
+__device__ thrust::pair<parse_result, json_output> get_json_object_single(
+  char const* input,
+  size_t input_len,
+  path_operator const* const commands,
+  char* out_buf,
+  size_t out_buf_size)
+{
+  json_state j_state(input, input_len);
+  json_output output{out_buf_size, 0, false, out_buf};
+
+  auto const result = parse_json_path<max_command_stack_depth>(j_state, commands, output);
+
+  return {result, output};
+}
+
+/**
+ * @brief Kernel for running the JSONPath query.
+ *
+ * This kernel operates in a 2-pass way.  On the first pass, it computes
+ * output sizes.  On the second pass it fills in the provided output buffers
+ * (chars and validity)
+ *
+ * @param col Device view of the incoming string
+ * @param commands JSONPath command buffer
+ * @param output_offsets Buffer used to store the string offsets for the results of the query
+ * (nullptr in the size computation step)
+ * @param out_buf Buffer used to store the results of the query (nullptr in the size computation
+ * step)
+ * @param out_validity Output validity buffer (nullptr in the size computation step)
+ */
+__global__ void get_json_object_kernel(column_device_view col,
+                                       path_operator const* const commands,
+                                       size_type* output_offsets,
+                                       char* out_buf,
+                                       bitmask_type* out_validity)
```

Comment:
Similar to a `bool` that says "is this thing valid or not", a pointer that may or may not be `nullptr` is better replaced by an `optional` because that immediately tells the reader that this object may or may not have a value. With a pointer, you don't know if `nullptr` is a valid value without looking at the code/documentation. 

```suggestion
                                       thrust::optional<device_span<size_type>> output_offsets,
                                       thrust::optional<device_span<char>> out_buf,
                                       thrust::optional<device_span<bitmask_type>> out_validity)
```

Reply:
The `device_span` part may or may not be necessary since it seems like you don't care about the size of these buffers.

Reply:
Minor note:  output_offsets will always be valid here  (in the precompute step will fill it in as sizes and in the compute step they are fed back in as offsets).

Reply:
Also, why a span instead of just pointers here?  Kind of makes the code ugly:

`char* dst  = out_buf.has_value() ? out_buf.value().data()+ output_offsets[tid] : nullptr;`